BLE: Add privacy trace

pan- opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Add traces to the privacy controller and its associated PAL. 
Traces of the host resolution operation are set at the debug level to not clutter the output.  

Preceding PR: https://github.com/ARMmbed/mbed-os/pull/14118

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [x] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

pan- committed:
WIP: SM Bluetooth traces

pan- committed:
BLE: Move traces helper into their own file

pan- committed:
BLE: Move traces helper into their own file

pan- committed:
BLE: Add traces to Security Manager PAL

pan- committed:
BLE: Make SM traces consistent.
- Prefix with Connection <id> - when appropriate.
- Display parameters after `:`.
- If multiple parameters should be displayed name then and print the value after =. They are separated by a `,`.

pan- committed:
BLE: Fix SM random number generation.

Only the first 8 bytes were generated from the stack.

pan- committed:
BLE: Regenerate CSRK if it is all zeroes.

pan- committed:
BLE: Add trace into privacy modules

Address resolution is set at the debug level as it is a very common operation and may clutter the output.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalPrivateAddressControllerImpl.cpp:
```diff
@@ -106,8 +120,12 @@ ble_error_t PalPrivateAddressController::add_device_to_resolving_list(
     }
+
+    tr_info("Add RPA to LL resolving list: peer address=%s, type=%s, peer irk=%s, local irk=%s",
+        to_string(peer_address_type), to_string(peer_identity_address), to_string(peer_irk), to_string(local_irk));
```
address and type swapped

```suggestion
        to_string(peer_identity_address), to_string(peer_address_type), to_string(peer_irk), to_string(local_irk));
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalPrivateAddressControllerImpl.cpp:
```diff
@@ -130,9 +148,12 @@ ble_error_t PalPrivateAddressController::remove_device_from_resolving_list(
     }
 
+    tr_info("Remove RPA from LL resolving list: peer address=%s, type=%s",
+        to_string(peer_address_type), to_string(peer_identity_address));
```
```suggestion
        to_string(peer_identity_address), to_string(peer_address_type));
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalPrivateAddressControllerImpl.cpp:
```diff
@@ -19,24 +19,32 @@
+#include "mbed-trace/mbed_trace.h"
+#include "common/ble_trace_helpers.h"
+
+#define TRACE_GROUP "BLSM"
```
maybe it's worth using a different tag: BLPR?

pan- replied:
Agreed.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalSecurityManagerImpl.cpp:
```diff
@@ -444,6 +482,7 @@ ble_error_t PalSecurityManager::cancel_pairing(
 ble_error_t PalSecurityManager::get_random_data(byte_array_t<8> &random_data)
 {
     SecRand(random_data.data(), random_data.size());
+    tr_info("Stack generated random data: %s", to_string(random_data));
```
don't think this is useful, if the data is relevant the caller will log it

pan- replied:
That still useful, the caller will log things like `set CSRK: .....` It doesn't say much about the origin of the bytes. 
I'd keep it, it help me in identifying the issue fixed by https://github.com/ARMmbed/mbed-os/pull/14125

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalSecurityManagerImpl.cpp:
```diff
@@ -697,6 +752,7 @@ bool PalSecurityManager::sm_handler(const wsfMsgHdr_t *msg)
         }
 
         case DM_SEC_LTK_REQ_IND: {
+            tr_info("Handling event DM_SEC_LTK_REQ_IND");
```
traces stop here, but there are a few more events below

pan- replied:
Thanks, I don't know what I was thinking.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/PrivateAddressController.cpp:
```diff
@@ -340,6 +370,7 @@ ble_error_t PrivateAddressController::resolve_address_on_host(
 
 void PrivateAddressController::on_resolving_list_action_complete()
 {
+    tr_info("Resolving list action conpleted");
```
```suggestion
    tr_info("Resolving list action completed");
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -302,9 +346,14 @@ ble_error_t SecurityManager::requestPairing(connection_handle_t connection)
-    if (!_db) return BLE_ERROR_INITIALIZATION_INCOMPLETE;
+    tr_info("Connection %d - Accept pairing request", connection);
+    if (!_db) {
+        tr_error("Failure, db not initialized");
```
So it can be collapsed with the rest of them
```suggestion
        tr_error("Failure, DB not initialized");
```

paul-szczepanek-arm replied:
I guess same below, some copy pasting would solve this to avoid accidental differences in lines.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -1199,11 +1445,13 @@ void SecurityManager::enable_encryption_cb(
 
     SecurityDistributionFlags_t* flags = _db->get_distribution_flags(cb->db_entry);
     if (!flags) {
+        tr_error("Connection %d - Cannot enable encryption, no distribution flags found for entry %p", cb->connection, cb->db_entry);
```
honestly, if you're fixing things, this whole if statement should just be an `MBED_ASSERT(flags);`

paul-szczepanek-arm replied:
Same below and other instances

pan- replied:
The fix would be to validate the DB when it is loaded. I'm not doing it now.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -1716,31 +2043,38 @@ void SecurityManager::on_passkey_display(
+    PasskeyAscii ascii_passkey(passkey);
+    tr_info("Connection %d - Received passkey display %.*s",
+        connection, ascii_passkey.PASSKEY_LEN, ascii_passkey.value());
+    eventHandler->passkeyDisplay(connection, ascii_passkey.value());
```
should we null check eventHandler? you added the check to other places

paul-szczepanek-arm replied:
and another one below

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.cpp:
```diff
@@ -1763,27 +2101,38 @@ void SecurityManager::on_keypress_notification(
+    tr_info("Connection %d - Keypress notification %s received", connection, to_string(keypress));
     set_mitm_performed(connection);
+    if (!eventHandler) {
+        tr_error("Impossible to forward Keypress notification to application, no event handler registered");
```
the call below proceeds, needs to be in else block

paul-szczepanek-arm replied:
same below

paul-szczepanek-arm reviewed (approved):
we should just merge and I'll make a separate PR to fix the issues listed above

pan- committed:
Address trace review for SM:

- Put privacy traces in BLPR
- Add missing traces in PALSecurityManagerImpl.cpp
- Add missing EventHandler null pointer check
- Typo and parameters order fix.
